### PR TITLE
Keep docstring information in native version

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Generate proper __doc__ member for module and functions

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
@decabyte : this one is for you too!

**Comment by decabyte** [comment]
Awesome. I'll try it on a clean install. ;)

--
Valerio De Carolis - http://decabyte.it

**Comment by serge-sans-paille** [comment]
On Mon, Jul 27, 2015 at 01:07:53AM -0700, Valerio De Carolis wrote:
> Awesome. I'll try it on a clean install. ;)

If you can manage to do it on the given branch, before it gets merged,
that would be great!

File: pythran/cxxgen.py [hunk 1]
```diff
@@ -471,32 +472,50 @@ def generate(self):
...
+        self.docstrings = docstrings or {}
+
+        moduledoc = self.docstrings.get(None, "")
+        if moduledoc:
+            self.add_to_init(
+                Statement('boost::python::scope().attr("__doc__") = ' +
+                          self.docstring(moduledoc)))
+
+        self.add_to_preamble(
+            Include("boost/python/docstring_options.hpp"))
```
Comment (pbrunet) [inline pythran/cxxgen.py]: os path join?

File: pythran/toolchain.py [hunk 2]
```diff
@@ -152,21 +152,22 @@ def __str__(self):
...
-        mod.add_to_preamble([Include("pythonic/python/core.hpp")])
-        mod.add_to_preamble([Line("#ifdef _OPENMP\n#include <omp.h>\n#endif")])
-        mod.add_to_preamble(map(Include, _extract_specs_dependencies(specs)))
-        mod.add_to_preamble(content.body)
-        mod.add_to_init([
-            Line('#ifdef PYTHONIC_TYPES_NDARRAY_HPP\nimport_array()\n#endif')])
+        mod.add_to_preamble(Define("BOOST_PYTHON_MAX_ARITY", max_arity),
+                            Define("BOOST_SIMD_NO_STRICT_ALIASING", "1"),
+                            Include("pythonic/core.hpp"),
+                            Include("pythonic/python/core.hpp"),
```
Comment (pbrunet) [inline pythran/toolchain.py]: path join?

File: pythran/transformations/extract_doc_strings.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        pass
+    '''
+
+    def __init__(self):
+        super(ExtractDocStrings, self).__init__()
+        self.docstrings = dict()
+
+    def run(self, node, ctx):
+        super(ExtractDocStrings, self).run(node, ctx)
+        return self.docstrings
```
Comment (pbrunet) [inline pythran/transformations/extract_doc_strings.py]: WHy not using the self.result?

**Comment by pbrunet** [comment]
We removed docstring because it was to costly when we timed it. Is it better now that is it done a import time and not at function call time?

**Comment by decabyte** [comment]
Hello guys, tested this morning but not getting any docstring passed down to the compiled module. Are you adding the docs to both modules and functions?

This is the small example I'm testing against:

```python
"""This is an example module to be converted with pythran"""

#pythran export dprod(int list, int list)
def dprod(l0, l1):
	"""dprod calculates the product between two lists"""
	return sum(x * y for x, y in zip(l0, l1))
```

Btw why not update the README once this is feature has been completed?

**Comment by serge-sans-paille** [comment]
On Mon, Jul 27, 2015 at 03:03:31AM -0700, pbrunet wrote:
> We removed docstring because it was to costly when we timed it. Is it better now that is it done a import time and not at function call time?

yes that's done only once.

**Comment by serge-sans-paille** [comment]
On Mon, Jul 27, 2015 at 03:16:37AM -0700, Valerio De Carolis wrote:
> Hello guys, tested this morning but not getting any docstring passed down to the compiled module. Are you adding the docs to both modules and functions?
> 
> This is the small example I'm testing against:
> 
> ```python
> """This is an example module to be converted with pythran"""
> 
> #pythran export dprod(int list, int list)
> def dprod(l0, l1):
> 	"""dprod calculates the product between two lists"""
> 	return sum(x * y for x, y in zip(l0, l1))
> ```
> 
> Btw why not update the README once this is feature has been completed?

Using the correct branch? your example works fine with me.

File: pythran/cxxgen.py (hunk 1 above)
Comment (serge-sans-paille) [inline pythran/cxxgen.py]: > os path join?

Not needed for windows portability, / is accepted for include path

File: pythran/transformations/extract_doc_strings.py (hunk 3 above)
Comment (serge-sans-paille) [inline pythran/transformations/extract_doc_strings.py]: > WHy not using the self.result?

it's not an analysis (it prunes some nodes from tha AST) we did the same
for NormalizeIdentifiers

**Comment by decabyte** [comment]
Sorry my fault. Working perfectly with the right branch. :+1: 

```
Help on module example:

NAME
    example - This is a test module with some awesome functions!

FILE
    /tmp/example.so

FUNCTIONS
    dprod(...)
        dprod calculates the product between two lists
```

**Comment by serge-sans-paille** [comment]
@pbrunet : travis likes this PR !

**Pull request merged**